### PR TITLE
Alerting: Rule API

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
This PR is a preparation for implementing support for fine-grained access control in rule API. The current implementation of the rule API allow users to perform many opertaions in one request (create, update or delete a rule). The API model is processed by the data layer and therefore it is not possible to perform authorization in the controller.
In this PR I try to separate responsibilities between controller and data layer:
- Add validation logic for API models. It copies what's done in storage layer with some additional validations. The result of validation is either user-friendly error or a collection of AlertRule. This allows us to validate model before opening the transaction.
- Update data store's method UpsertRules to not populate the new rule from the existing rule. This is not necessary because API layer expects complete model of the rule. Otherwise it will fail validation.
- Added function that pre-calculates changes before updating the database. The result of the function is a collection of rules that needs to be added, updated and removed. Using these calculations I will add authorization logic in future PR. 
- Moved transaction out of storage layer. Changes made in https://github.com/grafana/grafana/pull/44939 allowed reusing session and abstract from Sql client. 
- The refactoring made possible to perform more correct quota validation. It is moved to the end of the transaction to consider state after all changes. 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by gotjosh** [comment]
@papagian can you take a look first? You have more context on this than I do.

File: pkg/services/ngalert/api/api_ruler.go [hunk 1]
```diff
@@ -243,62 +249,73 @@ func (srv RulerSrv) RoutePostNameRulesConfig(c *models.ReqContext, ruleGroupConf
...
-		// that is acceptable under the assumption that there will be only one alert rule under the rule group
-		// alternatively we should check the quotas after the rule group update
-		// and rollback the transaction in case of violation
-		limitReached, err := srv.QuotaService.QuotaReached(c, "alert_rule")
-		if err != nil {
-			return ErrResp(http.StatusInternalServerError, err, "failed to get quota")
-		}
-		if limitReached {
-			return ErrResp(http.StatusForbidden, errors.New("quota reached"), "")
+		if len(changes.Upsert) > 0 {
```
Comment (papagian) [inline pkg/services/ngalert/api/api_ruler.go:283]: any reason that you don't check `changes.newRules > 0` instead? if all rules in `changes.Upsert` are updates (no new rules) then I think there is no need to check the quota.

File: pkg/services/ngalert/api/api_ruler.go [hunk 2]
```diff
@@ -243,62 +249,73 @@ func (srv RulerSrv) RoutePostNameRulesConfig(c *models.ReqContext, ruleGroupConf
...
+		if err != nil {
+			return fmt.Errorf("failed to add or update rules: %w", err)
 		}
-		if r.GrafanaManagedAlert.UID != "" {
-			_, ok := alertRuleUIDs[r.GrafanaManagedAlert.UID]
-			if ok {
-				return ErrResp(http.StatusBadRequest, fmt.Errorf("conflicting UID %q found", r.GrafanaManagedAlert.UID), "failed to validate alert rule %q", r.GrafanaManagedAlert.Title)
+
+		for _, rule := range changes.Delete {
+			if err = srv.store.DeleteAlertRuleByUID(tranCtx, c.SignedInUser.OrgId, rule.UID); err != nil {
```
Comment (papagian) [inline pkg/services/ngalert/api/api_ruler.go:278]: nit: we could delete all the rules in one SQL query

File: pkg/services/ngalert/ngalert.go [hunk 3]
```diff
@@ -94,9 +94,9 @@ func (ng *AlertNG) init() error {
 
 	baseInterval := ng.Cfg.AlertingBaseInterval
 	if baseInterval <= 0 {
-		baseInterval = defaultBaseIntervalSeconds
+		baseInterval = defaultBaseIntervalSeconds * time.Second
+		ng.Cfg.AlertingBaseInterval = baseInterval
```
Comment (papagian) [inline pkg/services/ngalert/ngalert.go:98]: nit, `AlertingBaseInterval` should better move under `Cfg.UnifiedAlerting`.
I have noticed also that we never `ng.Cfg.AlertingBaseInterval` so it always gets the default one.

**Review by papagian — APPROVED** [review]
nice! I wonder whether we still need `DBStore.validateAlertRule()`

File: pkg/services/ngalert/ngalert.go (hunk 3 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/ngalert.go:98]: Yes. this is a hack until https://github.com/grafana/grafana/pull/45107 is  merged. I would also appreciate your review of it.

File: pkg/services/ngalert/api/api_ruler.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/api/api_ruler.go:279]: Makes sense! I will address that in the future PR

File: pkg/services/ngalert/api/api_ruler.go (hunk 1 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/api/api_ruler.go]: fixed.

**Review by JohnnyQQQQ — APPROVED** [review]
👍

**Pull request merged**